ecdsa/src/recovery.rs

```diff
@@ -285,7 +285,17 @@ where
     ) -> Result<Self> {
         let (r, s) = signature.split_scalars();
         let z = <Scalar<C> as Reduce<C::Uint>>::reduce_bytes(&bits2field::<C>(prehash)?);
-        let R = AffinePoint::<C>::decompress(&r.to_repr(), u8::from(recovery_id.is_y_odd()).into());
+
+        let mut r_bytes = r.to_repr();
+        if recovery_id.is_x_reduced() {
+            let restored = C::Uint::decode_field_bytes(&r_bytes).checked_add(&C::ORDER);
+            // No reduction should happen here if r was reduced
+            if restored.is_none().into() {
+                return Err(Error::new());
+            }
+            r_bytes = restored.unwrap().encode_field_bytes();
```

Comment:
You can avoid the `unwrap` with something like:

```suggestion
            match C::Uint::decode_field_bytes(&r_bytes).checked_add(&C::ORDER).into() {
                Some(restored) => {
                    r_bytes = restored.unwrap().encode_field_bytes();
                }
                None => {
                    // No reduction should happen here if r was reduced
                    return Err(Error::new());
                }
             }
```

Reply:
That's a good idea! Have done something along those lines in https://github.com/RustCrypto/signatures/pull/680/commits/2730286b91d3fb6b70c1ec235d7c5294c049f96c.